pkg/controller/seed-controller-manager/kubernetes/resources.go

```diff
@@ -265,7 +265,12 @@ func GetDeploymentCreators(data *resources.TemplateData, enableAPIserverOIDCAuth
 	if data.Cluster().Annotations[kubermaticv1.AnnotationNameClusterAutoscalerEnabled] != "" {
 		deployments = append(deployments, clusterautoscaler.DeploymentCreator(data))
 	}
-	if flag := data.Cluster().Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider]; flag {
+	// If CCM migration is ongoing defer the deployment of the CCM to the
+	// moment in which cloud controllers or the full in-tree cloud provider
+	// have been deactivated.
+	if data.Cluster().Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
+		(!metav1.HasAnnotation(data.Cluster().ObjectMeta, kubermaticv1.CCMMigrationNeededAnnotation) ||
+			data.KCMCloudControllersDeactivated()) {
```

Comment:
How's this going to behave in the following scenario:
* The ExternalCloudProvider feature is enabled (the first part of the `if` is `true`)
* There is no `CCMMigrationNeeded` annotation (this will make the whole `if` block `true`)
* The in-tree controllers are still not deactivated (`data.KCMCloudControllersDeactivated()` returns `false`)

In that case, the condition will be satisfied and we would deploy the external CCM, even if the in-tree plugins are still enabled.

I'm not sure if this is something that can actually happen, but if yes, then we might need to reconsider this condition.

I guess that if we implement the mutating webhook, the webhook will edit the Cluster object and apply the annotation before the seed-controller-manager will reconcile it. In that case, this condition wouldn't be satisfied (because the annotation is there) and we'll not deploy the external CCM. However, the question is how we should handle it until we don't implement the webhook?

Reply:
I see your point, but consider that at the moment if the `externalCloudProvider`  feature is added to an existing cluster a hard migration is triggered anyway so this PR is not worsening the situation at least :-)